File: 2019W12/PreppindataW12.py
import pandas as pd
from datetime import timedelta

xls = pd.ExcelFile("E:\PD12 - System errors.xlsx")

autolog = pd.read_excel(xls,"Automatic Error log") 
manlog = pd.read_excel(xls,"Manual capture error list") 

for k in ['Start','End']:
    manlog[k+'DateTime'] = manlog.apply(lambda x: pd.to_datetime(x[k+' Date'].strftime('%Y-%m-%d')+' '+ x[k+' Time'].strftime('%H:%M:%S')),axis=1)
    autolog[k+'DateTime'] = autolog.apply(lambda x: x[k+' Date / Time']+timedelta(hours=0.5)*(1 if k=='End' else -1),axis=1)
    manlog.drop([k+' Date',k+' Time'],axis=1,inplace = True)

df_dup = pd.merge(manlog, autolog, on='System',suffixes=('_man', '_auto'))
df_dup = df_dup.query('StartDateTime_auto <= StartDateTime_man and EndDateTime_auto >= EndDateTime_man') 

final = pd.merge(autolog, df_dup, how='left', on=['System','Start Date / Time','End Date / Time'])
final = final[['System','Error','Start Date / Time','End Date / Time','StartDateTime_man','EndDateTime_man']].copy()

final = pd.merge(final,manlog, how='outer', left_on=['System','StartDateTime_man','EndDateTime_man'], \
                 right_on=['System','StartDateTime','EndDateTime'] ,suffixes=('_a', '_m'), indicator='Check')

for k in ['Start','End']:
    final[k+' Date / Time'] = final.apply(lambda x: x[k+'DateTime'] if pd.isnull(x[k+' Date / Time']) else x[k+' Date / Time'],axis=1)
    final.drop([k+'DateTime',k+'DateTime_man'],axis=1,inplace = True)
final['Error'] = final.apply(lambda x: x['Error_m'] if pd.isnull(x['Error_a']) else x['Error_a'],axis=1)
final['Error source'] = final.apply(lambda x: 'Manual capture error list' if x['Check']=='right_only' else 'Automatic Error log',axis=1)
final['Downtime in Hours'] = final.apply(lambda x: (x['End Date / Time']-x['Start Date / Time'])/timedelta(minutes=60),axis=1)    
final['Total Downtime in Hours'] = final['Downtime in Hours'].groupby(final['System']).transform('sum')
final['% of system downtime'] = final['Downtime in Hours']/final['Total Downtime in Hours']  
final.drop(['Error_m','Error_a','Check'],axis=1,inplace = True)

File: 2019W12/preppindataw12.R
library(readxl)
library(dplyr)
library(fuzzyjoin)
library(lubridate)

autolog <- read_excel("E:/PD12 - System errors.xlsx", 
                      sheet = "Automatic Error log", col_types = c("date", 
                                                                   "date", "text", "text")) %>%
  mutate('StartDateTime'= `Start Date / Time` %m+% minutes(-30),
         'EndDateTime'= `End Date / Time` %m+% minutes(30))
  
manlog <- read_excel("E:/PD12 - System errors.xlsx", 
                     sheet = "Manual capture error list", 
                     col_types = c("date", "date", "date", 
                                   "date", "text", "text")) %>%
  mutate('StartDateTime'= as_datetime(paste(strftime(`Start Date`, format="%Y-%m-%d"),
                                        strftime(`Start Time`, format="%H:%M:%S", tz = "UTC"))),
         'EndDateTime'= as_datetime(paste(strftime(`End Date`, format="%Y-%m-%d"),
                                            strftime(`End Time`, format="%H:%M:%S", tz = "UTC")))) %>%
  select(c('System','Error','StartDateTime','EndDateTime'))


final <- fuzzy_full_join(autolog, manlog, 
                        by = c('System','StartDateTime','EndDateTime'),
                        match_fun = list(`==`, `<=`, `>=`)
                        ) %>%
  mutate('Start Date / Time'= as_datetime(ifelse(is.na(`Start Date / Time`),`StartDateTime.y`,`Start Date / Time`)),
         'End Date / Time'= as_datetime(ifelse(is.na(`End Date / Time`),`EndDateTime.y`,`End Date / Time`)),
         'System' = ifelse(is.na(`System.x`),`System.y`,`System.x`),
         'Error Source' = ifelse(is.na(`System.x`),'Manual capture error list','Automatic Error log'),
         'Error'= ifelse(is.na(`Error.x`),`Error.y`,`Error.x`)) %>%
  select(c('System','Error','Error Source','Start Date / Time','End Date / Time')) %>%
  mutate('Downtime in Hours'= as.numeric((`End Date / Time` - `Start Date / Time`)/60)) %>%
  group_by(`System`) %>%
  mutate('Total Downtime in Hours' = sum(`Downtime in Hours`),
         '% of system downtime'= `Downtime in Hours` / sum(`Downtime in Hours`))
  
View(final)
